### PR TITLE
storage: assert around snapshot sending/receiving

### PR DESCRIPTION
**Pull request by irfansharif**
In v2.1 log entries are shipped alongside snapshots. The log entries
included in snapshots (which also include the truncated state and the
applied index) cover all indexes in the range
[truncated-state.index + 1, applied-state]. We simply assert that this
is always the case. We also assert during log truncations that the
number of deleted entries is no more than what we expect (last index -
first index).

Additionally rename PendingPreemptiveSnapshotIndex to
PendingSnapshotIndex, as it applies to both raft snapshots and
pre-emptive snapshots.

Release note: None

### CODE REVIEW

**Comment by irfansharif** [comment]
@tbg: I'm likely missing other spots where asserts would be useful. If you know of any I missed, let me know.

**Review by tbg — COMMENTED** [review]
I can't think of other places to put these. There's some more information that we'll generally want to put into these errors plus we'll want to create RocksDB checkpoints whenever they occur, but other than that this looks good!

File: pkg/storage/batcheval/cmd_truncate_log.go [hunk 1]
```diff
@@ -103,10 +103,15 @@ func TruncateLog(
...
 	} else {
-		if _, _, _, err := engine.MVCCDeleteRange(ctx, batch, &ms, start.Key, end.Key, math.MaxInt64, /* max */
-			hlc.Timestamp{}, nil /* txn */, false /* returnKeys */); err != nil {
+		_, _, numDeleted, err := engine.MVCCDeleteRange(ctx, batch, &ms, start.Key, end.Key, math.MaxInt64, /* max */
+			hlc.Timestamp{}, nil /* txn */, false /* returnKeys */)
+		if err != nil {
 			return result.Result{}, err
 		}
+		if expNumDeleted := (args.Index - firstIndex); uint64(numDeleted) > expNumDeleted {
+			log.Fatalf(ctx, "expected to delete up to %d log entries, deleted %d entries",
```
[the remark this answers is not in the document]
Comment (irfansharif) [reply pkg/storage/batcheval/cmd_truncate_log.go]: Done.

File: pkg/storage/raft_log_queue.go [hunk 2]
```diff
@@ -365,8 +365,8 @@ func computeTruncateDecision(input truncateDecisionInput) truncateDecision {
 	// about to be added to the range (or is in Raft recovery). We don't want to
 	// truncate the log in a way that will require that new replica to be caught
 	// up via yet another Raft snapshot.
-	if input.PendingPreemptiveSnapshotIndex > 0 && decision.NewFirstIndex > input.PendingPreemptiveSnapshotIndex {
-		decision.NewFirstIndex = input.PendingPreemptiveSnapshotIndex
+	if input.PendingSnapshotIndex > 0 && decision.NewFirstIndex > input.PendingSnapshotIndex {
```
[the remark this answers is not in the document]
Comment (irfansharif) [reply pkg/storage/raft_log_queue.go]: Done (I'll send it for master instead).

File: pkg/storage/raft_log_queue.go [hunk 3]
```diff
@@ -351,8 +351,8 @@ func computeTruncateDecision(input truncateDecisionInput) truncateDecision {
 		// snapshots with overlapping bounds that put significant stress on the
 		// Raft snapshot queue.
 		if progress.State == raft.ProgressStateProbe {
-			if decision.NewFirstIndex > decision.Input.FirstIndex {
-				decision.NewFirstIndex = decision.Input.FirstIndex
+			if decision.NewFirstIndex > input.FirstIndex {
```
[the remark this answers is not in the document]
Comment (irfansharif) [reply pkg/storage/raft_log_queue.go]: Done (but we don't modify decision.Input, nor does it seem that it should, which is why I put this up).

File: pkg/storage/replica_raftstorage.go [hunk 4]
```diff
@@ -936,6 +936,12 @@ func (r *Replica) applySnapshot(
 			s.RaftAppliedIndex, snap.Metadata.Index)
 	}
 
+	if expLen := (s.RaftAppliedIndex - s.TruncatedState.Index); expLen != uint64(len(logEntries)) {
+		log.Fatalf(ctx,
+			"received inconsistent number of log entries: got %d entries, expected %d entries",
```
[the remark this answers is not in the document]
Comment (irfansharif) [reply pkg/storage/replica_raftstorage.go]: CreateCheckpoint is not available in v2.1, it was only added in https://github.com/cockroachdb/cockroach/pull/36867. 

> Print the raft applied index, truncated state, HardState and actual range ([a..b]) of indexes read.

Done.

File: pkg/storage/store_snapshot.go [hunk 5]
```diff
@@ -139,14 +139,19 @@ func (kvSS *kvBatchSnapshotStrategy) Receive(
...
 				SnapUUID:   snapUUID,
 				Batches:    batches,
 				LogEntries: logEntries,
 				State:      &header.State,
 				snapType:   snapTypeRaft,
 			}
+
+			expLen := inSnap.State.RaftAppliedIndex - inSnap.State.TruncatedState.Index
+			if expLen != uint64(len(logEntries)) {
+				log.Fatal(ctx, "missing log entries in snapshot", expLen, len(logEntries))
```
[the remark this answers is not in the document]
Comment (irfansharif) [reply pkg/storage/store_snapshot.go]: Instead of materializing the actual range of indexes received, and plumbing in HardState here (kvBatchSnapshotStrategy shouldn't know about that anyway), I'm warning loudly here instead. It will still Fatal and capture HardState + range of log indexes received when applying the snapshot.

File: pkg/storage/raft_log_queue.go [hunk 6]
```diff
@@ -388,17 +388,28 @@ func computeTruncateDecision(input truncateDecisionInput) truncateDecision {
...
 	// in a no-op).
-	if decision.NewFirstIndex < decision.Input.FirstIndex {
-		decision.NewFirstIndex = decision.Input.FirstIndex
+	if decision.NewFirstIndex < input.FirstIndex {
+		decision.NewFirstIndex = input.FirstIndex
 		decision.ChosenVia = truncatableIndexChosenViaFirstIndex
 	}
+
+	// Invariants: NewFirstIndex >= FirstIndex
+	//             NewFirstIndex <= LastIndex (if != 10)
```
[the remark this answers is not in the document]
Comment (irfansharif) [reply pkg/storage/raft_log_queue.go]: > I certainly wouldn't want it to return a truncate decision that has NewFirstIndex > LastIndex

This already does happen (and why my first revision failed teamcity). `input.FirstIndex` is set to [TruncatedState.Index + 1](https://github.com/cockroachdb/cockroach/blob/9d62feaf2122872c5584ed0ff4ba1dafe88e7cb6/pkg/storage/replica_raftstorage.go#L347), so 11 for uinit'ed replicas, whereas LastIndex is 10.

> The code in line 384 above seems like it would set NewFirstIndex := 10 in this case. Is that not what happens?

It [does](https://github.com/cockroachdb/cockroach/blob/a2948d0be761db5940fc67d420eb52f1efef7c99/pkg/storage/raft_log_queue.go#L384-L387), but given 10 < 11 (input.FirstIndex), it's [brought back up](https://github.com/cockroachdb/cockroach/blob/a2948d0be761db5940fc67d420eb52f1efef7c99/pkg/storage/raft_log_queue.go#L389-L394) to 11. And thus we have NewFirstIndex > LastIndex. So this is funky, only for uninit'ed replicas can we have input.FirstIndex > input.LastIndex (tnput.FirstIndex = input.LastIndex + 1).

**Review by tbg — APPROVED** [review]
:lgtm: so far, thanks! Curious if we can get CreateCheckpoint onto release-2.1. It's really our biggest punch we can land if this bug comes up again

> Reviewed 2 of 7 files at r1, 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42011#-:-LsS7ORq2T5gny-z3-3g:b-pl3lwv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @irfansharif and @tbg)

___
*[pkg/storage/raft_log_queue.go, line 354 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42011#-LsObUCW7f3P3lh3YwLA-r1-354:-LsS5R0k8TS_dxXgKsmO:b-6krh69) ([raw file](https://github.com/cockroachdb/cockroach/blob/60ec8cef2aa5b981ad6c0f4f6e2de7498c6fc025/pkg/storage/raft_log_queue.go#L354)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Done (but we don't modify decision.Input, nor does it seem that it should, which is why I put this up).
</blockquote></details>

I know, just de-risking the change regardless. As a reviewer, I can't tell whether`Input` is embedded in `truncateDecision`. If, for example, `ChosenVia` were really in `Input`, we'd have introduced some weirdness. I know nothing like that happened, but I like to keep the diffs braindead on release branches.
___
*[pkg/storage/raft_log_queue.go, line 397 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42011#-LsObUCW7f3P3lh3YwLA-r1-397:-LsS7KlH6L2EGJH45NKK:bw53o7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/60ec8cef2aa5b981ad6c0f4f6e2de7498c6fc025/pkg/storage/raft_log_queue.go#L397)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

> I certainly wouldn't want it to return a truncate decision that has NewFirstIndex > LastIndex

This already does happen (and why my first revision failed teamcity). `input.FirstIndex` is set to [TruncatedState.Index + 1](https://github.com/cockroachdb/cockroach/blob/9d62feaf2122872c5584ed0ff4ba1dafe88e7cb6/pkg/storage/replica_raftstorage.go#L347), so 11 for uinit'ed replicas, whereas LastIndex is 10.

> The code in line 384 above seems like it would set NewFirstIndex := 10 in this case. Is that not what happens?

It [does](https://github.com/cockroachdb/cockroach/blob/a2948d0be761db5940fc67d420eb52f1efef7c99/pkg/storage/raft_log_queue.go#L384-L387), but given 10 < 11 (input.FirstIndex), it's [brought back up](https://github.com/cockroachdb/cockroach/blob/a2948d0be761db5940fc67d420eb52f1efef7c99/pkg/storage/raft_log_queue.go#L389-L394) to 11. And thus we have NewFirstIndex > LastIndex. So this is funky, only for uninit'ed replicas can we have input.FirstIndex > input.LastIndex (input.FirstIndex = input.LastIndex + 1).
</blockquote></details>

Add that in a comment, please.
___
*[pkg/storage/replica_raftstorage.go, line 941 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42011#-LsObUCW7f3P3lh3YwLD-r1-941:-LsS6-dNB91RfkaJgGu0:b-3w0p7u) ([raw file](https://github.com/cockroachdb/cockroach/blob/60ec8cef2aa5b981ad6c0f4f6e2de7498c6fc025/pkg/storage/replica_raftstorage.go#L941)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

CreateCheckpoint is not available in v2.1, it was only added in https://github.com/cockroachdb/cockroach/pull/36867. 

> Print the raft applied index, truncated state, HardState and actual range ([a..b]) of indexes read.

Done.
</blockquote></details>

Ah, that's a real bummer because some deployments tend to auto-restart crashing nodes which will wipe the evidence. Does c1d8a2e4824ff87478561828dabc9fcf42626f09 backport to release-2.1 somewhat cleanly?
___
*[pkg/storage/replica_raftstorage.go, line 944 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42011#-LsS5uE2CbaAVOCuvjsh:-LsS5uE2CbaAVOCuvjsi:b-8bj8kn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab8683d402129271ea98d03497854b80a4b7c042/pkg/storage/replica_raftstorage.go#L944)):*
> ```Go
> 				"(RaftAppliedIndex=%d, TruncatedState.Index=%d, HardState=%s, ReceivedLogEntries=[%d,%d])",
> 			len(logEntries), expLen, s.RaftAppliedIndex, s.TruncatedState.Index,
> 			hs.String(), logEntries[0].Index, logEntries[len(logEntries)-1].Index)
> ```

logEntries will be empty if we see the same bug again, so make sure the assertion doesn't panic in that case.
___
*[pkg/storage/store_snapshot.go, line 283 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42011#-LsS6mHZBK1g64fiVpIP:-LsS6mHZBK1g64fiVpIQ:bgy93q7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab8683d402129271ea98d03497854b80a4b7c042/pkg/storage/store_snapshot.go#L283)):*
> ```Go
> 	// snapshot) and the truncated index should equal the number of log entries
> 	// shipped over.
> 	expLen := endIndex - firstIndex
> ```

This is the assertion that we expect to fire - the snapshot that was *sent* had zero entries. Make sure this gets an engine snapshot if you find that CreateCheckpoint does backport well enough. Without the snapshot I think we'll be unlikely to have much evidence left by the time we get to take a look. Like in the other place, also print the actual range of indexes we got here.

**Review by irfansharif — COMMENTED** [review]
TFTR. I'll try backporting CreateCheckpoint in a separate PR (also it looks like it's missing from 19.1). I'll ping back here if successful :crossed_fingers: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42011#-:-LsT9vch8yk9mTUE6lk5:bzcb1p7)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @tbg)

___
*[pkg/storage/raft_log_queue.go, line 354 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42011#-LsObUCW7f3P3lh3YwLA-r1-354:-LsTAB5vAdLzsFfbvURG:b-xujghg) ([raw file](https://github.com/cockroachdb/cockroach/blob/60ec8cef2aa5b981ad6c0f4f6e2de7498c6fc025/pkg/storage/raft_log_queue.go#L354)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I know, just de-risking the change regardless. As a reviewer, I can't tell whether`Input` is embedded in `truncateDecision`. If, for example, `ChosenVia` were really in `Input`, we'd have introduced some weirdness. I know nothing like that happened, but I like to keep the diffs braindead on release branches.
</blockquote></details>

Gotcha, that makes sense. I'll keep this heuristic in mind going forward.
___
*[pkg/storage/raft_log_queue.go, line 397 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42011#-LsObUCW7f3P3lh3YwLA-r1-397:-LsTAH_31DvxJ7M57PM3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/60ec8cef2aa5b981ad6c0f4f6e2de7498c6fc025/pkg/storage/raft_log_queue.go#L397)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Add that in a comment, please.
</blockquote></details>

Done.
___
*[pkg/storage/replica_raftstorage.go, line 941 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42011#-LsObUCW7f3P3lh3YwLD-r1-941:-LsTAItpDuvI4y8EUdVC:bmhdvh7) ([raw file](https://github.com/cockroachdb/cockroach/blob/60ec8cef2aa5b981ad6c0f4f6e2de7498c6fc025/pkg/storage/replica_raftstorage.go#L941)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Ah, that's a real bummer because some deployments tend to auto-restart crashing nodes which will wipe the evidence. Does c1d8a2e4824ff87478561828dabc9fcf42626f09 backport to release-2.1 somewhat cleanly?
</blockquote></details>

Trying this in a separate PR.
___
*[pkg/storage/replica_raftstorage.go, line 944 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42011#-LsS5uE2CbaAVOCuvjsh:-LsTANUH4DFMfR6c9neb:b-39v3cp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab8683d402129271ea98d03497854b80a4b7c042/pkg/storage/replica_raftstorage.go#L944)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

logEntries will be empty if we see the same bug again, so make sure the assertion doesn't panic in that case.
</blockquote></details>

Whoops, fixed.
___
*[pkg/storage/store_snapshot.go, line 283 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42011#-LsS6mHZBK1g64fiVpIP:-LsTAQib6mEIc-9rNErF:bmhdvh7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab8683d402129271ea98d03497854b80a4b7c042/pkg/storage/store_snapshot.go#L283)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This is the assertion that we expect to fire - the snapshot that was *sent* had zero entries. Make sure this gets an engine snapshot if you find that CreateCheckpoint does backport well enough. Without the snapshot I think we'll be unlikely to have much evidence left by the time we get to take a look. Like in the other place, also print the actual range of indexes we got here.
</blockquote></details>

Trying this in a separate PR.

**Review by tbg — APPROVED** [review]
> Reviewed 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42011)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Comment by irfansharif** [comment]
Rebased atop #42042, PTA(brief)L.

**Comment by irfansharif** [comment]
Ignore TC failures, it's due to #42083.

File: build/go-version-check.sh [hunk 7]
```diff
@@ -24,8 +24,8 @@ if (( version_major != 1 )) || (( version_minor < 10 )); then
   exit 1
 fi
 
-# Pending resolution of #35637
-if [ $version_minor -ge 12 ]; then
-  echo "go 1.12+ is known to produce invalid crdb builds, see https://github.com/cockroachdb/cockroach/issues/35637" >&2
-  exit 1
-fi
+# # Pending resolution of #35637
```
Comment (tbg) [inline build/go-version-check.sh]: Is this kosher? Is there anything we needed to fix to get this working that we haven't backported?
Also, please put this into a separate commit, with a proper explanation.

**Review by tbg — CHANGES_REQUESTED** [review]
Minor comments, but I'm confused about the go version check.

> Reviewed 5 of 5 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42011#-:-Lsv_u4U25l1Pd3pvKYe:b-yuw2gj)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif)

___
*[pkg/storage/replica_command.go, line 1033 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/42011#-Lsv_HP25dr7W4ljeHdk:-Lsv_HP25dr7W4ljeHdl:b-15yav3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3feb097f4668c33416f36bc7b2b4971be2fab862/pkg/storage/replica_command.go#L1033)):*
> ```Go
> 		sent,
> 	); err != nil {
> 		if _, ok := err.(*MalformedSnapshotError); ok {
> ```

This looks brittle, I'd go for `errors.Cause` to make sure that an intermittent `errors.Wrap` isn't letting the error bypass this check. Also just test it manually (by always returning this error) and making sure that any test that involves snapshots fails with the proper fatal.

**Review by irfansharif — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42011)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @tbg)

___
*[pkg/storage/replica_command.go, line 1033 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/42011#-Lsv_HP25dr7W4ljeHdk:-LswLwpO7Ykjcint2MfQ:bhwsvwc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3feb097f4668c33416f36bc7b2b4971be2fab862/pkg/storage/replica_command.go#L1033)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This looks brittle, I'd go for `errors.Cause` to make sure that an intermittent `errors.Wrap` isn't letting the error bypass this check. Also just test it manually (by always returning this error) and making sure that any test that involves snapshots fails with the proper fatal.
</blockquote></details>

Done. Also checked manually that checkpoints are created (tests create in mem rocksdb instances, which don't create checkpoints it seems).

**Review by tbg — APPROVED** [review]
> Reviewed 3 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42011)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Comment by tbg** [comment]
</god merge>
